roc/roc.go

```diff
@@ -1,170 +1,142 @@
-	__v := (int32)(__ret)
-	return __v
-}
-
-// receiverClose function as declared in roc/receiver.h:195
-func receiverClose(receiver *Receiver) int32 {
-	creceiver, _ := (*C.roc_receiver)(unsafe.Pointer(receiver)), cgoAllocsUnknown
-	__ret := C.roc_receiver_close(creceiver)
-	__v := (int32)(__ret)
-	return __v
-}
+	// ErrRuntime indicates a runtime error: memory allocation error etc
+	ErrRuntime = errors.New("Runtime error")
+)
 
-// senderOpen function as declared in roc/sender.h:111
-func senderOpen(context *Context, config *SenderConfig) *Sender {
-	ccontext, _ := (*C.roc_context)(unsafe.Pointer(context)), cgoAllocsUnknown
-	cconfig, _ := config.PassRef()
-	__ret := C.roc_sender_open(ccontext, cconfig)
-	__v := *(**Sender)(unsafe.Pointer(&__ret))
-	return __v
+// safeString ensures that the string is NULL-terminated, a NULL-terminated copy is created otherwise.
+func safeString(str string) string {
+	if len(str) > 0 && str[len(str)-1] != '\x00' {
+		str = str + "\x00"
+	} else if len(str) == 0 {
+		str = "\x00"
+	}
+	return str
 }
 
-// senderBind function as declared in roc/sender.h:134
-func senderBind(sender *Sender, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_bind(csender, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// ============================================================================
+//                              address.h
+// ============================================================================
 
-// senderConnect function as declared in roc/sender.h:153
-func senderConnect(sender *Sender, kind PortType, proto Protocol, address *Address) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	ckind, _ := (C.roc_port_type)(kind), cgoAllocsUnknown
-	cproto, _ := (C.roc_protocol)(proto), cgoAllocsUnknown
-	caddress, _ := address.PassRef()
-	__ret := C.roc_sender_connect(csender, ckind, cproto, caddress)
-	__v := (int32)(__ret)
-	return __v
-}
+// NewAddress parses the `ip`, `port` and `family` and initializes the Address object
+func NewAddress(family Family, ip string, port int) (*Address, error) {
+	a := new(Address)
+	a.mem = make([]byte, C.sizeof_roc_address)
+	a.raw = (*C.roc_address)(unsafe.Pointer(&a.mem))
 
-// senderWrite function as declared in roc/sender.h:177
-func senderWrite(sender *Sender, frame *frame) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	cframe, _ := frame.PassRef()
-	__ret := C.roc_sender_write(csender, cframe)
-	__v := (int32)(__ret)
-	return __v
-}
-
-// senderClose function as declared in roc/sender.h:192
-func senderClose(sender *Sender) int32 {
-	csender, _ := (*C.roc_sender)(unsafe.Pointer(sender)), cgoAllocsUnknown
-	__ret := C.roc_sender_close(csender)
-	__v := (int32)(__ret)
-	return __v
+	cfamily := (C.roc_family)(family)
+	ip = safeString(ip)
+	cip := C.CString(ip)
```

Comment:
```
// Go string to C string
// The C string is allocated in the C heap using malloc.
// It is the caller's responsibility to arrange for it to be
// freed, such as by calling C.free (be sure to include stdlib.h
// if C.free is needed).
func C.CString(string) *C.char
```

There is a leak here because CString performs malloc. Actually we don't need CString, we can just cast `ip` to `C.char*`.

Reply:
I don't think we can do it, I tried this:
`cip := (*C.char)(unsafe.Pointer(&ip))` and the test showed that address is nil (`NewAdress` failed)

what we can do is:
1) allocate and then free
2) make `[]C.char` of fixed size and copy byte-by-byte

Reply:
> cip := (*C.char)(unsafe.Pointer(&ip))

This doesn't work because `&ip` is the address of the *string object*, but not the address of the string data. (string object is a pointer to string data + its length, IIRC).

However, I just checked and we really can not cast string to `*C.char` because in Go you can't get an address of the string data (i.e. `&s[0]` is a compilation error).

But you can cast it to []byte and then cast (though it may cost an allocation):

```
$ cat 1.go 
package main

/*
#include <stdio.h>

void foo(const char* str) {
        printf("<%s>\n", str);
}
*/
import "C"
import (
        "runtime"
        "unsafe"
)

func Foo(s string) {
        b := []byte(s)
        C.foo((*C.char)(unsafe.Pointer(&b[0])))
        runtime.KeepAlive(b)
}

func main() {
        Foo("hello")
}
```

```
$ go run 1.go
<hello>
```

> allocate and then free
> make []C.char of fixed size and copy byte-by-byte

Agree, also good.